### PR TITLE
Enabling the cancellation of non-actor tasks in a worker's queue 2

### PR DESCRIPTION
**Pull request by goliaro**
## Why are these changes needed?
This reopens PR #12117 after it was reverted

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/test/mock_worker.cc [hunk 1]
```diff
@@ -79,6 +79,8 @@ class MockWorker {
     } else if ("MergeInputArgsAsOutput" == typed_descriptor->ModuleName()) {
       // Merge input args and write the merged content to each of return ids
       return MergeInputArgsAsOutput(args, return_ids, results);
+    } else if ("WhileTrueLoop" == typed_descriptor->ModuleName()) {
+      return WhileTrueLoop(args, return_ids, results);
```
Comment (ffbin) [inline src/ray/core_worker/test/mock_worker.cc:83]: Will this line of code be executed?

Comment (gabrieleoliaro) [reply src/ray/core_worker/test/mock_worker.cc:83]: It should be executed when we run the TestCancelTasks C++ test (in the core_worker_test.cc)

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -2238,6 +2238,9 @@ void CoreWorker::HandleCancelTask(const rpc::CancelTaskRequest &request,
   reply->set_attempt_succeeded(success);
   send_reply_callback(Status::OK(), nullptr, nullptr);
 
+  // TODO: fix race condition to avoid using this hack
+  requested_task_running = main_thread_task_id_ == task_id;
+
```
Comment (ijrsvt) [inline src/ray/core_worker/core_worker.cc:2243]: Maybe set this before the `send_reply_callback`?

Comment (gabrieleoliaro) [reply src/ray/core_worker/core_worker.cc:2243]: ok, let me try that!

File: src/ray/core_worker/test/mock_worker.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/core_worker/test/mock_worker.cc:83]: Hmm, `WhileTrueLoop` is a dead loop. How can the program exit?

File: src/ray/core_worker/test/mock_worker.cc [hunk 3]
```diff
@@ -128,6 +130,14 @@ class MockWorker {
     return Status::OK();
   }
 
+  Status WhileTrueLoop(const std::vector<std::shared_ptr<RayObject>> &args,
+                       const std::vector<ObjectID> &return_ids,
+                       std::vector<std::shared_ptr<RayObject>> *results) {
+    while (1) {
+    }
```
Comment (ffbin) [inline src/ray/core_worker/test/mock_worker.cc:137]: How about add a sleep to reduce CPU load?

Comment (gabrieleoliaro) [reply src/ray/core_worker/test/mock_worker.cc:137]: that's a good idea! I should add that

**Review by stephanie-wang — APPROVED** [review]

**Pull request merged**